### PR TITLE
Filter app

### PR DESCRIPTION
**Pull request by danyoungday**
Redid app to use our new flow of selecting a context -> filtering prescriptors by desired outcome -> choosing a prescriptor -> play with the actions -> see the outcomes.

Additionally updated deployment to download a smaller, preprocessed dataset instead of the entire thing. We also save the demo models in the repo for simplicity.

### CODE REVIEW

File: use_cases/eluc/app/app.py [hunk 1]
```diff
@@ -1,118 +1,49 @@
...
+app_df = pd.read_csv("app/data/app_data.csv", index_col=app_constants.INDEX_COLS)
 
-trivia_component = TriviaComponent(df)
-trivia_component.register_update_trivia_callback(app)
+evolution_handler = EvolutionHandler()
 
+intro_component = IntroComponent()
+context_component = ContextComponent(app_df, evolution_handler)
+filter_component = FilterComponent(evolution_handler)
+dms_component = DMSComponent(app_df, evolution_handler)
```
Comment (danyoungday) [inline use_cases/eluc/app/app.py:31]: We pass our app df and evolution handler to the components so that they have the ability to prescribe, look at context, etc.

File: use_cases/eluc/app/app.py [hunk 2]
```diff
@@ -1,118 +1,49 @@
...
-This site is for demonstration purposes only.
-
-For a given context cell representing a portion of the earth,
-identified by its latitude and longitude coordinates, and a given year:
-* What changes can we make to the land usage
-* In order to minimize the resulting estimated CO2 emissions? (Emissions from Land Use Change, ELUC,
- in tons of carbon per hectare)
+context_component.register_callbacks(app)
+filter_component.register_callbacks(app)
+dms_component.register_callbacks(app)
```
Comment (danyoungday) [inline use_cases/eluc/app/app.py:36]: Instead of all the long series of callbacks we now register all the callbacks at once with these simple functions. We could abstract these into a general `Component` class later and just call `register_callbacks` on them all.

File: use_cases/eluc/app/app.py [hunk 3]
```diff
@@ -1,118 +1,49 @@
...
-], style={'padding-left': '10px'},)
+app.layout = html.Div(
+    children=[
+        intro_component.get_div(),
+        context_component.get_div(),
+        filter_component.get_div(),
+        dms_component.get_div(),
+        references_component.get_references_div()
+    ]
+)
```
Comment (danyoungday) [inline use_cases/eluc/app/app.py:46]: Layout is now super clean. Just calls get_div from each of our components in the order they should display on the apge

File: use_cases/eluc/app/assets/tooltip.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+window.dccFunctions = window.dccFunctions || {};
+window.dccFunctions.percentSlider = function(value) {
+     return Math.round(value * 100);
+}
```
Comment (danyoungday) [inline use_cases/eluc/app/assets/tooltip.js:4]: So that our sliders show as percents and not proportions (10% vs. 0.1)

File: use_cases/eluc/app/components/context/context.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+from app.utils import EvolutionHandler
+from data import constants
+
+
+class ContextComponent():
+    """
+    Component containing map as well as dropdowns and input fields for picking a more specific context.
+    """
+    def __init__(self, app_df: pd.DataFrame, handler: EvolutionHandler):
+        self.map_component = MapComponent(app_df)
```
Comment (danyoungday) [inline use_cases/eluc/app/components/context/context.py:19]: We reuse the old map component in our new context component

File: use_cases/eluc/app/components/context/context.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        self.app_df = app_df
+        self.handler = handler
+        self.countries_df = regionmask.defined_regions.natural_earth_v5_0_0.countries_110.to_dataframe()
+
+    def create_label_and_value(self, label: str, value: html.Div) -> html.Div:
+        """
+        Standard dash function that pairs a label with any arbitrary value Div.
+        """
```
Comment (danyoungday) [inline use_cases/eluc/app/components/context/context.py:27]: This function lines up a label with a div so we reuse it for the Lat/dropdown, lon/dropdown, year/input

File: use_cases/eluc/app/components/context/map.py
Comment (danyoungday) [inline use_cases/eluc/app/components/context/map.py:1]: Trimmed a lot of the fat off and only kept what we use in the new app

File: use_cases/eluc/app/components/dms/chart.py [hunk 7]
```diff
@@ -37,60 +38,61 @@ def register_update_context_chart_callback(self, app):
         """
         @app.callback(
             Output("context-fig", "figure"),
-            Input("chart-select", "value"),
             Input("year-input", "value"),
             Input("lat-dropdown", "value"),
             Input("lon-dropdown", "value")
         )
-        def update_context_chart(chart_type, year, lat, lon):
+        def update_context_chart(year, lat, lon):
```
Comment (danyoungday) [inline use_cases/eluc/app/components/dms/chart.py:45]: We don't use the pie charts anymore but may add them back in later

File: use_cases/eluc/app/components/dms/chart.py [hunk 8]
```diff
@@ -37,60 +38,61 @@ def register_update_context_chart_callback(self, app):
...
+            :return: New chart using presc data.
             """
-            # If we have no prescription just return an empty chart
-            if all(slider == 0 for slider in sliders):
-                return self.create_treemap(pd.Series([], dtype=float), type_context=False, year=year)
 
-            presc = pd.Series(sliders, index=constants.RECO_COLS)
+            # If we have no prescription just show the context chart
+            bad_sliders = sum(sliders) < -0.01 or sum(sliders) > 0.01
+            if all(slider == 0 for slider in sliders) or bad_sliders:
```
Comment (danyoungday) [inline use_cases/eluc/app/components/dms/chart.py:84]: We add a `prescribe + predict` button instead of automatically updating the chart so we don't have to validate that the sliders sum to 100 on the fly

File: use_cases/eluc/app/components/dms/dms.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+            slider_vals = selected[constants.DIFF_RECO_COLS].iloc[0].tolist()
+            min_vals = selected[constants.RECO_COLS].iloc[0].tolist()
+            min_vals = [-1 * min_val for min_val in min_vals]
+            return slider_vals, min_vals
+
+        @app.callback(
+            Output("pbar", "children"),
+            [Input({"type": "diff-slider", "index": ALL}, "value")]
+        )
+        def update_pbar(sliders) -> int:
```
Comment (danyoungday) [inline use_cases/eluc/app/components/dms/dms.py:159]: We use a progress bar to show how much land is used

File: use_cases/eluc/app/components/filter.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+import app.constants as app_constants
+from app.utils import EvolutionHandler
+from data import constants
+
+
+class FilterComponent:
+    """
+    Component with sliders/preset buttons to filter outcomes by as well as visualizing the selected prescriptors.
+    """
```
Comment (danyoungday) [inline use_cases/eluc/app/components/filter.py:18]: This is almost the exact same as the climate change demo

File: use_cases/eluc/app/components/filter.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+    def __init__(self, handler: EvolutionHandler):
+        self.handler = handler
+        self.slider_params = ["min", "max", "value", "marks"]
+
+    def plot_prescriptions(self, results_df: pd.DataFrame, cand_ids: list[str]) -> go.Figure:
+        """
+        Plots the prescriptions for the selected candidates and grays the rest out, shifting them to the right.
+        Also adds the original context as the far leftmost bar.
+        """
```
Comment (danyoungday) [inline use_cases/eluc/app/components/filter.py:27]: Logic for our nice stacked bar chart

File: use_cases/eluc/app/components/filter.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        @app.callback(
+            Output({"type": "filter-slider", "index": ALL}, "value", allow_duplicate=True),
+            Input({"type": "preset-button", "index": ALL}, "n_clicks"),
+            State("results-store", "data"),
+            prevent_inital_call=True
+        )
+        def select_preset(n_clicks, results_json):
+            """
+            Selects a preset for the filter sliders from the 3 presets.
+            """
```
Comment (danyoungday) [inline use_cases/eluc/app/components/filter.py:200]: We make the presets just slice the outcomes into thirds and each third is our preset

File: use_cases/eluc/app/requirements.txt [hunk 13]
```diff
@@ -1,4 +1,4 @@
-dash==2.10.2
+dash==2.15
```
Comment (danyoungday) [inline use_cases/eluc/app/requirements.txt:1]: Needed to update dash to get functionality to have the sliders show as percentages

File: use_cases/eluc/app/process_data.py [hunk 14]
```diff
@@ -1,25 +1,11 @@
...
-    test_df.to_csv(save_dir / "app_data.csv")
+from huggingface_hub import hf_hub_download
 
 
 if __name__ == "__main__":
-    main()
+    hf_hub_download(repo_id="projectresilience/land-use-app-data",
+                    filename="app_data.csv",
+                    local_dir="app/data",
+                    repo_type="dataset")
```
Comment (danyoungday) [inline use_cases/eluc/app/process_data.py:11]: Our process data script no longer downloads the entire dataset then processes it. It simply downloads a preprocessed file I left in huggingface

File: use_cases/eluc/app/utils.py [hunk 15]
```diff
@@ -1,21 +1,116 @@
...
+from prescriptors.prescriptor_manager import PrescriptorManager
+from prescriptors.nsga2.candidate import Candidate
+from prescriptors.nsga2.land_use_prescriptor import LandUsePrescriptor
+from predictors.percent_change.percent_change_predictor import PercentChangePredictor
+
+
+def add_nonland(context_actions_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Adds a nonland column to the context_actions_df
+    """
```
Comment (danyoungday) [inline use_cases/eluc/app/utils.py:24]: This function is modified to take a dataframe now instead of a series

File: use_cases/eluc/app/utils.py [hunk 16]
```diff
@@ -1,21 +1,116 @@
...
+
+        candidate_params = {"in_size": len(constants.CAO_MAPPING["context"]),
+                            "hidden_size": 16,
+                            "out_size": len(constants.RECO_COLS)}
+        prescriptors = {}
+        for cand_id in self.prescriptor_list:
+            cand_path = self.prescriptor_path / f"{cand_id}.pt"
+            candidate = Candidate(**candidate_params, cand_id=cand_id)
+            candidate.load_state_dict(torch.load(cand_path))
+            prescriptors[cand_id] = LandUsePrescriptor(candidate, encoder)
```
Comment (danyoungday) [inline use_cases/eluc/app/utils.py:65]: For now the prescriptors are hard-coded and saved in the repo. In the future these should live in HuggingFace

File: use_cases/eluc/requirements.txt [hunk 17]
```diff
@@ -8,7 +8,7 @@ dask==2023.6.1
 datasets==2.16.1
 flake8==7.1.0
 Flask==2.2.5
-geopandas==0.13.2
+geopandas==0.14.4
```
Comment (danyoungday) [inline use_cases/eluc/requirements.txt:11]: Updated geopandas because we were getting an error with the `fiona` package

File: use_cases/eluc/requirements.txt [hunk 18]
```diff
@@ -20,7 +20,7 @@ pandas==1.5.3
 plotly==5.14.1
 protobuf==3.20.3
 -e git+https://github.com/Project-Resilience/sdk.git@c0e5300e56f5fce2f6cf55c328c83e7055e541be#egg=prsdk
-pylint==2.17.4
+pylint==3.3.1
```
Comment (danyoungday) [inline use_cases/eluc/requirements.txt:23]: Updated pylint as we were linting unnecessary files

File: use_cases/eluc/app/app.py [hunk 19]
```diff
@@ -1,118 +1,49 @@
...
 
 if __name__ == '__main__':
-    app.run_server(host='0.0.0.0', debug=False, port=4057, use_reloader=False, threaded=False)
+    app.run_server(host='0.0.0.0', debug=False, port=4057, use_reloader=True, threaded=True)
```
Comment (ofrancon) [inline use_cases/eluc/app/app.py]: Did you mean to leave user_reloader to True?

Comment (ofrancon) [reply use_cases/eluc/app/app.py]: According to https://dash.plotly.com/devtools you're not changing the code in the deployed app, and it makes it slower to start, so I'll leave it to False

**Review by ofrancon — APPROVED** [review]
lgtm
